src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -650,6 +627,97 @@ int IEqualityComparer<ImmutableHashSet<string>>.GetHashCode(ImmutableHashSet<str
             return hash;
         }
 
+        private class DisplayNameToItemsMap : IEnumerable<CompletionItem>
+        {
+            // We might need to handle large amount of items with import completion enabled,
+            // so use a dedicated pool to minimize array allocations.
+            // Set the size of pool to a small number 5 because we don't expect more than a
+            // couple of callers at the same time.
+            private static readonly ObjectPool<Dictionary<string, object>> s_uniqueSourcesPool
+                = new(factory: () => new(), size: 5);
+
+            private readonly Dictionary<string, object> _displayNameToItemsMap;
+            private readonly CompletionServiceWithProviders _service;
+
+            public DisplayNameToItemsMap(CompletionServiceWithProviders service, int capacity)
+            {
+                _service = service;
+                _displayNameToItemsMap = s_uniqueSourcesPool.Allocate();
+#if NETCOREAPP
+                _displayNameToItemsMap.EnsureCapacity(capacity);
+#endif
+            }
+
+            public void Free()
+            {
+                _displayNameToItemsMap.Clear();
+                s_uniqueSourcesPool.Free(_displayNameToItemsMap);
+            }
+
+            public bool IsEmpty => _displayNameToItemsMap.Count == 0;
+
+            public void Add(CompletionItem item)
+            {
+                var entireDisplayText = item.GetEntireDisplayText();
+
+                if (!_displayNameToItemsMap.TryGetValue(entireDisplayText, out var value))
+                {
+                    _displayNameToItemsMap.Add(entireDisplayText, item);
+                    return;
+                }
+
+                // If two items have the same display text choose which one to keep.
+                // If they don't actually match keep both.
+                if (value is CompletionItem sameNamedItem)
+                {
+                    if (_service.ItemsMatch(item, sameNamedItem))
+                    {
+                        _displayNameToItemsMap[entireDisplayText] = _service.GetBetterItem(item, sameNamedItem);
+                        return;
+                    }
+
+                    _displayNameToItemsMap[entireDisplayText] = new List<CompletionItem>() { sameNamedItem, item };
+                }
+                else if (value is List<CompletionItem> sameNamedItems)
+                {
+                    for (var i = 0; i < sameNamedItems.Count; i++)
+                    {
+                        var existingItem = sameNamedItems[i];
+                        if (_service.ItemsMatch(item, existingItem))
+                        {
+                            sameNamedItems[i] = _service.GetBetterItem(item, existingItem);
+                            return;
+                        }
+                    }
+
+                    sameNamedItems.Add(item);
+                }
+            }
+
+            public IEnumerator<CompletionItem> GetEnumerator()
```

Comment:
do you actually enumerate this dictionary?  if so, you could consider making a struct enumrator so you save on this allocatoin as well.

Reply:
yes, but only once (per completion session), so it's probably no t necessary